src/prefect/tasks/secrets/vault_secret.py

```diff
@@ -0,0 +1,161 @@
+import re
+import json
+import os
+
+try:
+    import hvac
+except ImportError as err:
+    raise ImportError(
+        "Using `prefect.tasks.secrets.vault_secret` requires hvac to be installed."
+    ) from err
+
+import prefect
+from prefect.tasks.secrets import SecretBase, PrefectSecret
+from prefect.utilities.tasks import defaults_from_attrs
+from prefect.exceptions import ClientError
+
+
+class VaultSecret(SecretBase):
+    """
+    Vault Secrets Task.  This task retrieves a secret using the Hashicorp Vault service.
+    Local secrets are checked before querying the Vault service.
+
+    Client configuration:
+        - vault_addr: set via env var.  Both VAULT_ADDR and vault_addr supported
+
+    Args:
+        - name (str): The secret name defined by the Vault secret path "<mount-point>/<path>"
+        - vault_credentials_secret (str, optional): name of the PrefectSecret containing Vault
+            credentials.
+                Defaults to a PrefectSecret named `VAULT_CREDENTIALS`.
+                Supported vault client authentication methods:
+                * token { 'VAULT_TOKEN: '<token>' }
+                * appRole: { 'VAULT_ROLE_ID': '<role-id>',
+                             'VAULT_SECRET_ID': '<secret-id>' }
+        - **kwargs (Any, optional): additional keyword args passed to the Task constructor
+
+    Raises:
+        - ValueError: if a `result` keyword is passed
+    """
+
+    def __init__(
+        self, name: str, vault_credentials_secret: str = "VAULT_CREDENTIALS", **kwargs
+    ):
+        self.vault_credentials_secret = vault_credentials_secret
+        kwargs["name"] = name
+        super().__init__(**kwargs)
+
+    def _get_vault_secret(self, name: str):
+        """
+        Connect to vault service and retrieve a secret value
+        Arguments:
+            - name (str): name of the secret to be retrieved using "<mount_point>/<path>"
+
+        Returns:
+            - secret (dict): a dict of the secret key/value items retrieved
+
+        Raises:
+            - ClientError: unable to configure the vault client
+            - PermissionError: unable to authenticate or attempting an unsupported auth method
+            - KeyError: unable to lookup secret in vault using the secret path
+            - PermissionError: provided token/role not authorised to access the secret
+        """
+        self.logger.debug(f"looking up vault path: {name}")
+        client = hvac.Client()
+        # get vault address url
+        vault_url = os.getenv("VAULT_ADDR", default=None)
+        if vault_url is None:
+            vault_url = os.getenv("vault_addr", default=None)
+            if vault_url is None:
+                raise ClientError(
+                    "VAULT_ADDR url var not found. "
+                    'Either "VAULT_ADDR" or "vault_addr" env var required.'
+                )
```

Comment:
```suggestion
        vault_url = os.getenv("VAULT_ADDR") or os.getenv("vault_addr")
            raise ValueError(
                "VAULT_ADDR url var not found. "
                'Either "VAULT_ADDR" or "vault_addr" env var required.'
            )
```
- `os.getenv` defaults to `None` by default, I don't think it's worth being more explicit there
- `ClientError` is reserved for errors in the `prefect.Client`
- Is there precedence for checking for the lowercase variable as well? If that's not something vault generally does, just the upper-case should suffice

Reply:
I agree but believe you're missing a line in there before the raise? @madkinsz

Reply:
Thanks Tyler -- edited.

Reply:
Code simplified and `ClientError` exceptions and docs updated to use `ValueError`.

The check for both upper case and lower case versions enables support for 2 different methods for configuring env vars in the job execution environment.  Over time, we've used  both scenarios.
The Vault utilities expect an upper case `VAULT_ADDR` value.  If using a custom base image, the `ENV VAULT_ADDR=<vault-add>` env var can be set in the image definition and the value used by both prefect jobs and other scripts that use the same base image.
We also use a helm chart to configure the agent and job configuration, setting the vault address there, in case we have a flow that doesn't use our base image.  We use env vars in the chart to set value, which then get [interpolated to lower case keys](https://docs.prefect.io/core/concepts/configuration.html#environment-variables), e.g. from the deployment spec:
```yaml
...
    env:
    - name: PREFECT__CLOUD__AGENT__ENV_VARS__vault_addr
      value: "<vault-addr>"
...
```

Reply:
Thanks!

Hmm. I see your point regarding the lower-casing of environment variables. We explicitly don't do that with secrets and I think we may want to replicate that logic for the `AGENT__ENV_VARS__{VAR}` setting -- that seems like an oversight.